### PR TITLE
Checkboxes in the flyout

### PR DESCRIPTION
**Pull request by rachel-fenichel**
Open questions:
What event should be fired on click?  Currently it just logs the id of the block
What field on the block says that a checkbox is necessary?  Right now they're in front of statement blocks.

### CODE REVIEW

**Comment by tmickel** [comment]
> What event should be fired on click? Currently it just logs the id of the block

Maybe can we do a UI event that has the block ID and some name, similar to the category switch event?

> What field on the block says that a checkbox is necessary? Right now they're in front of statement blocks.

Something like `hasReporterCheckbox` perhaps?

File: core/css.js [hunk 1]
```diff
@@ -952,5 +952,9 @@ Blockly.Css.CONTENT = [
     'padding: 0;',
   '}',
 
+  '.blocklyFlyoutCheckbox {',
+    'fill: red',
+  '}',
```
Comment (tmickel) [inline core/css.js:957]: is the end goal to have a separate class for when it's checked vs. not checked?

File: core/flyout.js [hunk 2]
```diff
@@ -684,6 +720,44 @@ Blockly.Flyout.prototype.layoutBlocks_ = function(blocks, gaps) {
 };
 
 /**
+ * Create and place a checkbox corresponding to the given block.
+ * @param {!Blockly.Block} block The block to associate the checkbox to.
+ * @param {number} cursorX The x position of the cursor during this layout pass.
+ * @param {number} cursorY The y position of the cursor during this layout pass.
+ * @param {!{height: number, width: number}} blockHW The height and width of the
+ *     block.
+ * @param {Element} root The svg root of the block.
```
Comment (tmickel) [inline core/flyout.js]: why pass in these properties (blockHW, root) when you pass in `block`?

File: core/flyout.js [hunk 3]
```diff
@@ -684,6 +720,44 @@ Blockly.Flyout.prototype.layoutBlocks_ = function(blocks, gaps) {
...
+Blockly.Flyout.prototype.createCheckbox_ =
+    function(block, cursorX, cursorY, blockHW, root) {
+  /* eslint-disable indent */
+  var extraSpace = this.CHECKBOX_SIZE + this.CHECKBOX_MARGIN;
+  var checkboxRect = Blockly.createSvgElement('rect',
+    {
+      'class': 'blocklyFlyoutCheckbox',
+      'height': this.CHECKBOX_SIZE,
+      'width': this.CHECKBOX_SIZE,
+      'x': this.RTL ? cursorX + blockHW.width + extraSpace :
```
Comment (tmickel) [inline core/flyout.js:743]: Hmmm, something seems odd in RTL:
screen shot 2016-06-24 at 11 35 57 am
Thoughts?

File: core/flyout.js [hunk 4]
```diff
@@ -684,6 +720,44 @@ Blockly.Flyout.prototype.layoutBlocks_ = function(blocks, gaps) {
...
+ * @param {number} cursorX The x position of the cursor during this layout pass.
+ * @param {number} cursorY The y position of the cursor during this layout pass.
+ * @param {!{height: number, width: number}} blockHW The height and width of the
+ *     block.
+ * @param {Element} root The svg root of the block.
+ * @private
+ */
+Blockly.Flyout.prototype.createCheckbox_ =
+    function(block, cursorX, cursorY, blockHW, root) {
+  /* eslint-disable indent */
```
Comment (tmickel) [inline core/flyout.js]: I'm wondering why you choose to disable eslint vs. have some of the properties come on the next line.

File: core/flyout.js [hunk 5]
```diff
@@ -730,6 +810,11 @@ Blockly.Flyout.prototype.addBlockListeners_ = function(root, block, rect) {
       block.addSelect));
   this.listeners_.push(Blockly.bindEvent_(rect, 'mouseout', block,
       block.removeSelect));
+
+  if (block.flyoutCheckbox) {
+    this.listeners_.push(Blockly.bindEvent_(block.flyoutCheckbox, 'mousedown', null,
```
Comment (tmickel) [inline core/flyout.js]: touch events?

File: core/flyout.js [hunk 6]
```diff
@@ -130,6 +137,29 @@ Blockly.Flyout.prototype.MARGIN = 8;
 Blockly.Flyout.prototype.SCROLLBAR_PADDING = 2;
 
 /**
+ * Size of a checkbox next to a variable reporter.
+ * @type {number}
+ * @const
+ */
+Blockly.Flyout.prototype.CHECKBOX_SIZE = 20;
```
Comment (tmickel) [inline core/flyout.js:146]: Can this be set and calculated in CSS, or is it necessary for layout that it's defined in the JS? It comes to mind that we could want to have the checkbox a different size in touch environments from desktop environments...I wonder how we should deal with that if it comes up.

**Comment by tmickel** [comment]
Also wondering about the way to switch a checkbox to checked/unchecked from the outside. Do you want to save that for a separate PR?

File: core/flyout.js (hunk 3 above)
Comment (rachel-fenichel) [inline core/flyout.js:743]: Aha, that's probably what was stopping me from sending out this PR.  Will take a look.

File: core/flyout.js (hunk 5 above)
Comment (rachel-fenichel) [inline core/flyout.js]: What about them?  This is the same way that we add listeners on the blocks in the flyout.

File: core/flyout.js (hunk 4 above)
Comment (rachel-fenichel) [inline core/flyout.js]: Good point.  Will fix.

File: core/flyout.js (hunk 5 above)
Comment (tmickel) [inline core/flyout.js]: so you're saying it should just work? I haven't tried it, I just saw 'mousedown' and wondered :)

Comment (tmickel) [inline core/flyout.js]: yep, looks to me like bindEvent_ deals with that. sounds good

File: core/flyout.js (hunk 6 above)
Comment (rachel-fenichel) [inline core/flyout.js:146]: That sounds like a bigger question.  Leaving it this way for now.

File: core/flyout.js (hunk 2 above)
Comment (rachel-fenichel) [inline core/flyout.js]: GetHeightWidth() is not cheap--it walks the tree of blocks.  The results are needed here and in the caller function.

root has a simple getter.  I can get it in here instead.

**Comment by rachel-fenichel** [comment]
Setting a checkbox to checked or unchecked from the outside--I think the way to do it is to call flyout.setCheckboxState(blockId, state), which then 
1) gets the block by id and 
2) checks or unchecks the checkbox by calling flyout.checkboxClicked, setting the class [that I haven't created yet] on the associated checkbox.  The blocks have a reference so that's easy.

If that sounds reasonable, I can add it now.  Unclear when we'll have callers though.

**Comment by tmickel** [comment]
^ sounds great to me. If you want a way to test it, I guess we could add a button to the playground that randomly checks/unchecks the checkbox on all the blocks or something.

**Comment by rachel-fenichel** [comment]
Updated with setCheckboxState, better tracking of checkboxes and their blocks, and proper handling of RTL and zoom.

**Comment by tmickel** [comment]
Looks good to me, except it seems checking the checkbox causes the page to get text selected:
jun-27-2016 23-09-42

And I'd like to add the JSON property/option for setting the checkbox asap :)

**Comment by rachel-fenichel** [comment]
#260

**Pull request merged**